caput/mpiarray.py

```diff
@@ -1036,6 +1036,177 @@ def _partition_io(self, skip=False, threshold=1.99):
+        https://numpy.org/doc/stable/reference/ufuncs.html
+
+        ndarray has lots of built-in ufuncs. In order to use them, the MPIArrays need to be converted into ndarrays, otherwise NumPy reports a NotImplemented error.
+
+        The distributed axis for all input MPIArrays, is expected to be the same. Operations across the distributed axis, will not be permitted.
+
+        The new array will ehter be be distributed over that axis, or possibly one axis down for `reduce` methods, if they eliminate the distributed axis.
+
+        For operations that normally return a scalar, the scalars will be wrapped into a 1D array, distributed across axis 0.
+
+        Parameters
+        ----------
+        ufunc: <function>
+            ufunc object that was called
+        method: str
+            indicates which ufunc method was called. one of "__call__", "reduce", "reduceat", "accumulate", "outer", "inner"
+        inputs: tuple
+            tuple of the input arguments to the ufunc. At least one of the inputs is an MPIArray.
+        kwargs: dict
+            dictionary containing the optional input arguments of the ufunc. Important kwargs considered here are 'out' and 'axis'.
+        """
+        # pylint: disable=no-member
+        # known problem with super().__array_ufunc__
+
+        args = []
+
+        # convert all local arrays into ndarrays
+        args, dist_axis = MPIArray._mpi_to_ndarray(inputs)
+
+        if "axis" in kwargs and (kwargs["axis"] == dist_axis):
+            raise ValueError(
+                f"operations along the distributed axis (in this case, {dist_axis}) are not allowed."
+            )
+
+        # 'out' kwargs contain arrays that the ufunc places the results into
+        # this broadcasts the local part of the output arrays into an ndarray
+        # that the ufunc knows how to work with
+        outputs = kwargs.get("out", None)
+        if outputs:
+            out_args, _ = MPIArray._mpi_to_ndarray(outputs)
+            kwargs["out"] = tuple(out_args)
+        else:
+            outputs = (None,) * ufunc.nout
+
+        results = super().__array_ufunc__(ufunc, method, *args, **kwargs)
+
+        if results is NotImplemented:
+            return NotImplemented
+
+        # operation is performed in-place, so we can just return
+        if method == "at":
+            return
+
+        if ufunc.nout == 1:
+            results = (results,)
+
+        if "reduce" in method:
+            # reduction methods eliminate axes, so the distributed axes needs to be recalculated
+            if results[0].shape and len(results[0].shape) < dist_axis + 1:
+                dist_axis = len(results[0].shape) - 1
+
+        # Wrapping the results back into MPIArrays, distributed across the appropriate axis
+        ret = []
+
+        for result, output in zip(results, outputs):
+            if (
+                output is not None
+            ):  # case: results were placed in the array specified by `out`; return as is
```

Comment:
This and the clause below are where black seems to have made the code less readable. I would suggest just putting the comment within the block below the clause.